vitrc.5
.TH vitrc 5 "" "" "User Manuals"

.SH NAME
vitrc \- Configuration file for the vit(1) command

.SH SYNOPSIS
.B $HOME/.vitrc

.SH DESCRIPTION
.B VIT
obtains its configuration data from a file called
.I .vitrc
\&. This file must be located in the user's home directory.

Except for comments (lines starting with "#"), currently the only acceptable
commands in .vitrc are
.B map
commands and
.B set
commands.

.SS MAP COMMANDS
.B map
commands take the form

.RS
map <KEY>=<KEY-SEQUENCE>
.RE

where:
.RS
.TP
<KEY>
should represent a key or combination of keys, such as "a" or a <KEY-CODE> as
listed below.

.P
and
.P
<KEY-SEQUENCE>
is a sequence of <KEY>s to emulate. The result should be the same as if these
keys were pressed manually by the user.

<KEY-CODE>s are strings that represent keys. The following may be used:
.br
Function keys: <F1> <F2> <F3> <F4> <F5> <F6> <F7> <F8> <F9> <F10> <F11> <F12>

Navigation keys: <Home> <End> <PageUp> <PageDown> <Up> <Down> <Right> <Left>

Editing keys: <Insert> <Del> <Backspace> <Space> <Tab> <Return>

Modifier combinations: \\ca for Ctrl + "a" combination.

.SS SET COMMANDS
.B set
commands take the form

.RS
set <VAR NAME>=<VAR VAL>
.RE

where valid variable names (<VAR NAME>) and their acceptable values (<VAR VAL>)
are listed below. Note that the default value is listed immediately after the
variable name.

.TP
.B burndown=no
When set to "yes", VIT will display the time until convergence in the upper
left corner. It is set to "no" by default because it requires extra time to
obtain (from running 'task burndown' in the background. This variable was
essentially set to "yes" for VIT versions 1.2 and below.

.TP
.B confirmation=yes
When set to "yes" (default), VIT will ask for confirmation before marking a task as
done, deleting a task, or quitting VIT. Set to "no" to disable the prompts.

.TP
.B wait=yes
When set to "no", VIT will not show output of the task command after
modifications to a task are made.  By default, VIT will show the output of the
task command and wait for enter.

.SH EXAMPLES
.SS EXTERNAL COMMANDS
Note that for many of the examples, you need to have the appropriate extension
installed and in your PATH. Recall that after :! a 'w' means that VIT should
"wait". This allows another command to take over (e.g. Vim when you're editing
a task) and for you to see the output of a command. After :! a 'r' means that
VIT should reread your list of tasks. This should be used when your external
command changes the task data.

.TP
.B map \\\\cn=:!wr tasknote %TASKID<Return>
When Ctrl + n is pressed, runs the command 'tasknote <ID>', where <ID> is the
ID of the task that is currently selected in VIT.

.TP
.B map c=:! task %TASKID rc.defaultwidth:200 | grep -o -P -m 1 "(ftp|http)s?://\\\\S*" | tr -d '\n' | xsel -b<Return>
Copies the first link it finds to the clipboard.

.TP
.B map Z=:!rw echo "Changing due date\n"; task %TASKID modify due:$(read -p "How many days? " DAYS; echo $DAYS)days<Return>
Asks you for input and then changes the due date of the selected task to NOW plus that many days.

.TP
.B map <Left>=:!wr task $(task _get %TASKID.parent) edit<Return>
When the left arrow key is pressed, edits the parent of a recurring task
(do not use on non-recurring task).

.SS INTERNAL COMMANDS
.TP
.B map Q=
Disables whatever the Q key is bound to internally in VIT, which is a quick quit.


.SH "CREDITS & COPYRIGHTS"
.PP
Copyright (C) 2012 - 2013, Steve Rader
.br
Copyright (C) 2013 - 2019, Scott Kostyshak


VIT is distributed under GNU General Public License version 3 or (at your
option) any later version. See
http://www.gnu.org/licenses/ for more information.

.SH SEE ALSO
.BR vit(1)
.BR taskrc(1)

For more information regarding VIT, see the following:

.TP
The official site at
<http://tasktools.org/projects/vit.html>

.TP
The official code repository at
<https://git.tasktools.org/EX/vit.git>

.TP
You can contact the project by emailing
<support@taskwarrior.org>

.SH REPORTING BUGS
.TP
Bugs in VIT may be reported to the issue-tracker at
<http://taskwarrior.org>



CHANGES
 Sat Jun 14 2018 - fix "Negative repeat count does nothing" errors (GH#153)
 Sat Jun 24 2017 - introduce feature for adding/removing tags (GH#5)
 Sat Jun 24 2017 - add config option to disable wait prompts (GH#12)
 Fri Jun 23 2017 - add config option to disable confirmation (GH#4)
 Fri Jun 23 2017 - add "wait" command bound to 'w' (GH#3)
 Sun Mar 12 2017 - add ctrl+g synonym for escape
 Sun Mar 12 2017 - add start/stop toggle bound to 'b' (GH#152, part of GH#126)
 Sat Aug 30 2016 - fix handling of CJK characters (GH#142)
 Sat Aug  6 2016 - fix for multi-byte searches
 Sun Jul 24 2016 - fix for multi-byte prompt input
 Sun Jul 24 2016 - fix display of multi-byte report data
 Sun Jul 17 2016 - do not exit if terminal resized to small height
 Mon Jun 20 2016 - work around another instance of the Perl warning
 Mon May 16 2016 - work around a warning from Perl versions >= v5.21.1
 Sun Jan 10 2016 - annotations now correctly escape any character
 Sun Jan 10 2016 - do not exit when invalid regex for search string (GH#148)
 Sat Jan  9 2016 - clean up terminal on a Perl error or warning (part of GH#148)
 Fri Jan  8 2016 - do not run 'task burndown' by default
 Fri Jan  8 2016 - add config variable for 'task burndown'
 Tue Jan  5 2016 - add support for config variables. See 'man vitrc'
 Wed Sep 30 2015 - improve detection of annotations (GH#144)
 Sun Mar  8 2015 - do not beep on a 'g' keystroke
 Sat Mar  7 2015 - fix a bug where prompt text was invisible
 Wed Mar  4 2015 - add support for End and Home keys (GH#137)
 Mon Mar  2 2015 - add support for Del key in prompts (GH#120)
 Sun Mar  1 2015 - allow Taskwarrior cmds to parse for 'a' and 'm' (GH#132, GH#135)
 Sun Mar  1 2015 - in screen, VIT now highlights the entire line (GH#81, GH#129)
 Sun Mar  1 2015 - update display after sync (GH#112)
 Tue Aug 26 2014 - 'n'/'N' now work when not right after search
 Sun Aug 24 2014 - store commands file in %prefix%/share/vit/
 Wed Jul 30 2014 - update documentation URLs
 Fri Jun 27 2014 - add prompt history scrolling with arrows (GH#54, GH#58)
 Thu Jun 26 2014 - install files in more conventional paths (GH#118)
 Wed Jun 25 2014 - fix vitrc man file (GH#119)
 Wed Jun 25 2014 - Makefile no longer requires sudo (GH#118)
 Sat Jun 21 2014 - '-version' prints the git hash if available
 Fri Jun 20 2014 - all of vit's options work with two dashes
 Fri Jun 20 2014 - 'vit -help' has 0 exit code
 Fri Jun 20 2014 - 'vit -version' prints the version (GH#114)

 Sun Apr  6 2014 - released v1.2
 Tue Apr  1 2014 - the <Esc> key can now be used in shortcuts
 Tue Apr  1 2014 - exit with informative error if shortcut too long (see GH#103)
 Thu Mar 13 2014 - fix colors for running VIT in tmux
 Sat Mar  8 2014 - do not print control characters to prompts
 Thu Mar  6 2014 - fix recognition of backspace in tmux
 Thu Mar  6 2014 - fix a prompt bug that prevented editing
 Mon Mar  3 2014 - 'vit -audit' now creates a log with debug info

 Tue Feb  4 2014 - released v1.2.beta1
 Tue Feb  4 2014 - Add VIT man pages (#1284)
 Mon Oct 21 2013 - Implement cursor movement in prompts (#1403)
 Sun Oct  6 2013 - Clear project prompt string if escape (#1232)
 Sun Oct  6 2013 - Remove confusing behavior from arrow keys in prompts (#1363)
 Sun Sep 28 2013 - 'P' now sets priority and 'h', 'm', 'l', 'n' are freed (#1238)
 Sun Sep 27 2013 - 'c' is renamed to 'm' (#1231)
 Sun Sep 15 2013 - 't' now opens the command prompt with ":!rw task "
 Sun Sep 15 2013 - shell commands can now pass the arguments VIT is using (#1338, #1237)
 Sat Sep 14 2013 - custom keybinds can now be specified in ~/.vitrc (#1237, #1302, #1336)
 Thu Sep 12 2013 - added ':!' to execute arbitrary string in shell
 Sun Aug 12 2013 - When running an external command, VIT no longer echoes it
 Sun Aug 11 2013 - VIT now cleans the terminal before exiting
 Sun Aug 11 2013 - 'q' ('Q') now quits with(out) confirmation (#1266)
 Thu Aug  9 2013 - fix a bug where prompt text was invisible
 Mon Aug  5 2013 - 's' now runs 'task sync' if Taskwarrior >= 2.3.0 (#1301)
 Sun Aug  4 2013 - when in search mode, backspace now removes a character
 Sat Aug  3 2013 - 'D' now deletes a task when not over an annotation (#1230)
 Sat Jul  6 2013 - added Copyright 2013, Scott Kostyshak
 Sat Jul  6 2013 - added an AUTHORS file listing contributors
 Sat Jul  6 2013 - 'gg' now moves to first line (#1229)
 Sun Jun 23 2013 - added Copyright 2012 - 2013, Steve Rader

 Wed Apr  3 2013 - released v1.1
 Wed Apr  3 2013 - fixes for not having color=on set in ~/.taskrc
 Sun Mar 31 2013 - added <enter> for task info
 Sun Mar 31 2013 - added logging error msgs when "-audit" is used
 Sun Mar 31 2013 - added support for selection effects (e.g. bold)
 Sun Mar 31 2013 - added setting the VIT header color via color.vit.header setting
 Sat Mar 30 2013 - set the VIT header color via the color.header setting
 Fri Mar 29 2013 - added support for the "inverse" and "bright" effects
 Fri Mar 29 2013 - fixed parsing some ANSI underline effect escape sequences
 Fri Mar 29 2013 - clear the screen before exec'ing external commands as per feature #1214
 Fri Mar 29 2013 - fixed a bug where some commands (e.g. ":h") incorrectly waited after exec'ing
 Fri Mar 29 2013 - added setting the default report via command line args as per feature #1216
 Fri Mar 29 2013 - added support to allow for verbose=off as per topic #2851
 Fri Mar 29 2013 - disallowed using a default.command which doesn't include an "ID" column
 Fri Mar 29 2013 - added support for multiple effects, e.g. bold underline 
 
 Sun Mar 24 2013 - released v1.0
 Sun Mar 24 2013 - added '=' for task info as per feature #1156
 Sun Mar 24 2013 - added 'u' for task undo
 Thu Jan  1 2013 - fixed a bug where '/' and '?' caused a crash as per bug #1152
 Wed Dec 12 2012 - added graceful handling of marking only task in current report "done"
 Wed Dec 12 2012 - added "blinking" of the convergence info when convergence changes
 Wed Dec 12 2012 - disallowed marking completed tasks as "done"

vitrc.pl
# Copyright 2013 - 2019, Scott Kostyshak

sub parse_vitrc {
  my $vitrc = glob("~/.vitrc");
  if ( open(IN,"<$vitrc") ) {
    while (<IN>) {
      chop;
      my $parse_error = "ERROR: incorrect key bind line in .vitrc:\n $_\n";
      if ( $_ =~ s/^map // ) {
        my($scut, $cmd) = split(/=/, $_, 2);

        my $skey;
        if ($scut =~ s/([^ ]+)$//) {
          $skey = $1;
        }
        else {
          print STDERR "$parse_error";
          exit(1);
        }

        $skey = &replace_keycodes("$skey");
        $cmd = &replace_keycodes("$cmd");

        # TODO: get rid of the eval().
        # This shouldn't be too hard, but I need to figure out what exactly is happening here.
        $skey = eval "\"$skey\"";

        $shortcuts{$skey} = $cmd;
      }
      elsif ( $_ =~ s/^set // ) {
        my($configname, $configval) = split(/=/, $_, 2);
        &audit("CONFIG: user requests to set '$configname' to '$configval'");
        if ($configname eq "burndown") {
          # TODO: fix code duplication (see below)
          if (!&sanitycheck_bool($configval)) {
            print STDERR "ERROR: boolean config variable '$configname' must ".
                         "be set to 'yes' or 'no'.\n";
            exit(1);
          }
          $burndown = $configval;
        }
        elsif ($configname eq "confirmation") {
          # TODO: fix code duplication (see above)
          if (!&sanitycheck_bool($configval)) {
            print STDERR "ERROR: boolean config variable '$configname' must ".
                         "be set to 'yes' or 'no'.\n";
            exit(1);
          }
          $confirmation = ( $configval eq "yes" ? 1 : undef );
        }
        elsif ($configname eq "wait") {
          # TODO: more code duplication
          if (!&sanitycheck_bool($configval)) {
            print STDERR "ERROR: boolean config variable '$configname' must ".
                         "be set to 'yes' or 'no'.\n";
            exit(1);
          }
          $nowait = ($configval eq "no");
        }
      }
    }
    close(IN);
  }
}

#------------------------------------------------------------------

sub replace_keycodes {
  my $str_ = $_[0];

  $str_ =~ s/<F1>/KEY_F(1)/e;
  $str_ =~ s/<F2>/KEY_F(2)/e;
  $str_ =~ s/<F3>/KEY_F(3)/e;
  $str_ =~ s/<F4>/KEY_F(4)/e;
  $str_ =~ s/<F5>/KEY_F(5)/e;
  $str_ =~ s/<F6>/KEY_F(6)/e;
  $str_ =~ s/<F7>/KEY_F(7)/e;
  $str_ =~ s/<F8>/KEY_F(8)/e;
  $str_ =~ s/<F9>/KEY_F(9)/e;
  $str_ =~ s/<F10>/KEY_F(10)/e;
  $str_ =~ s/<F11>/KEY_F(11)/e;
  $str_ =~ s/<F12>/KEY_F(12)/e;

  $str_ =~ s/<Home>/KEY_HOME/e;
  $str_ =~ s/<End>/KEY_END/e;

  $str_ =~ s/<Insert>/KEY_IC/e;
  $str_ =~ s/<Del>/KEY_DC/e;

  $str_ =~ s/<PageUp>/KEY_PPAGE/e;
  $str_ =~ s/<PageDown>/KEY_NPAGE/e;

  $str_ =~ s/<Up>/KEY_UP/e;
  $str_ =~ s/<Down>/KEY_DOWN/e;
  $str_ =~ s/<Right>/KEY_RIGHT/e;
  $str_ =~ s/<Left>/KEY_LEFT/e;

  $str_ =~ s/<Backspace>/KEY_BACKSPACE/e;

  # We don't evaluate these ones (no 'e').
  $str_ =~ s/<Space>/ /;
  $str_ =~ s/<Tab>/\t/;
  $str_ =~ s/<Return>/\n/;
  $str_ =~ s/<Esc>/\e/;

  return $str_;
}

#------------------------------------------------------------------

sub sanitycheck_bool {
  my $bool_ = $_[0];
  if ($bool_ ne "yes" && $bool_ ne "no") {
    return 0;
  }
  else {
    return 1;
  }
}

return 1;

